apps/prairielearn/src/ee/lib/lti13.ts

```diff
@@ -211,3 +258,227 @@ export async function validateLti13CourseInstance(
+      type: 'external_tool',
+      external_tool_url: assessment.url,
+    },
+  };
+
+  job.info(
+    `Creating assignment for ${assessment.label} in ${instance.lti13_instance.name} ${instance.lti13_course_instance.context_label}`,
+  );
+
+  const token = await access_token(instance.lti13_instance.id);
+  const item: Lti13Lineitem = await fetchRetry(instance.lti13_course_instance.lineitems, {
+    method: 'POST',
+    headers: {
+      Authorization: `Bearer ${token}`,
+      'Content-type': 'application/vnd.ims.lis.v2.lineitem+json',
+    },
+    body: JSON.stringify(createBody),
+  });
+
+  job.info('Associating PrairieLearn assessment with the new assignment');
+
+  await link_assessment(instance.lti13_course_instance.id, assessment.id, item);
+  job.info('Done.');
+}
+
+export async function query_and_link_lineitem(
+  instance: any,
+  lineitem_id: string,
+  assessment_id: string | number,
+) {
+  const lineitems = await get_lineitems(instance);
+  const item = lineitems.find(({ id }) => id === lineitem_id);
+
+  if (item) {
+    await link_assessment(instance.lti13_course_instance.id, assessment_id, item);
+  } else {
+    throw new HttpStatusError(400, 'Lineitem not found');
+  }
+}
+
+export async function unlink_assessment(
+  lti13_course_instance_id: string,
+  assessment_id: string | number,
+) {
+  await queryAsync(sql.delete_lineitem_by_assessment_id, {
+    lti13_course_instance_id,
+    assessment_id,
+  });
+}
+
+export async function link_assessment(
+  lti13_course_instance_id: string,
+  assessment_id: string | number,
+  lineitem: Lti13Lineitem,
+) {
+  await queryAsync(sql.update_lineitem, {
+    lti13_course_instance_id,
+    lineitem_id: lineitem.id,
+    lineitem: JSON.stringify(lineitem),
+    assessment_id,
+  });
+}
+
+function sleep(delay: number) {
+  return new Promise((resolve) => setTimeout(resolve, delay));
+}
+
+/* Throttling notes
+// https://canvas.instructure.com/doc/api/file.throttling.html
+// 403 Forbidden (Rate Limit Exceeded)
+// X-Request-Cost
+// X-Rate-Limit-Remaining
+*/
+
+export async function fetchRetry(
+  input: RequestInfo | URL,
+  opts?: RequestInit | undefined,
+  incomingfetchRetryOpts?,
+) {
+  const fetchRetryOpts = {
+    retryLeft: 5,
+    sleepMs: 1000,
+    ...incomingfetchRetryOpts,
+  };
+  try {
+    const response = await fetch(input, opts);
+
+    if (!response.ok) {
+      throw new HttpStatusError(response.status, response.statusText);
+    }
+
+    const parsed = parseLinkHeader(response.headers.get('link')) ?? {};
+    if ('next' in parsed) {
+      const results = await response.json();
+
+      if (Array.isArray(results)) {
+        return results.concat(await fetchRetry(parsed.next.url, opts, fetchRetryOpts));
+      }
+    }
+
+    return await response.json();
+  } catch (err) {
+    fetchRetryOpts.retryLeft -= 1;
+    if (fetchRetryOpts.retryLeft === 0) {
+      throw err;
+    }
+    await sleep(fetchRetryOpts.sleepMs);
```

Comment:
Did you consider using exponential backoffs?

Reply:
I did, but the note at https://canvas.instructure.com/doc/api/file.throttling.html made me feel like a more complicated backoff would just be overly complex.

> Since the cost of a request is roughly based on the amount of time it takes to process, and the quota (by default) replenishes at a rate faster than real-time, any API client that makes no more than one simultaneous request is unlikely to be throttled.

We know we're likely to queue these kinds of commands anyway, so I didn't make this too complicated for a MVP.

Reply:
Famous last words 😉